File: ProjectEuler/Euler12.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProjectEuler
{
    /// <summary>
    /// What is the value of the first triangle number to have over five hundred divisors?
    /// </summary>
    public class Euler12
    {
        public static long FirstTriangleNumberWith500Divisors()
        {
            long valueWithMostDivisorsSoFar = 0;
            long numberOfDivisors = 0;
            long incrementor = (int)1e3;

            for (long i = 1; ; i += incrementor)
            {
                var num = TriangleNumberGen(i);
                var divisors = HasNumOfDivisors(num, 500);

                if (divisors > numberOfDivisors)
                {
                    numberOfDivisors = divisors;
                    valueWithMostDivisorsSoFar = num;

                    Console.WriteLine("Value: {0}, divisors: {1}", valueWithMostDivisorsSoFar, divisors);
                }

                if (divisors > 450) incrementor = 1;
                if (divisors > 500) break;
            }

            return valueWithMostDivisorsSoFar;
        }

        private static long TriangleNumberGen(long triangleNumberPosition)
        {
            long triangleNumber = 0;
            Parallel.For(1, triangleNumberPosition, x =>
            {
                triangleNumber += x;
            });

            return triangleNumber + triangleNumberPosition;
        }

        private static long HasNumOfDivisors(long triangleNum, long numberOfDivisors)
        {
            long divisors = 0;
            Parallel.For(1, triangleNum, x =>
            {
                if (!Euler7.IsPrime(triangleNum))
                {
                    if (triangleNum % x == 0)
                        divisors++;
                }
            });

            return divisors;
        }
    }
}

File: ProjectEulerFSharp/Euler12.fs
namespace Euler

open System.Collections.Generic
open System.Collections
open System
open System.Linq

// What is the value of the first triangle number to have over five hundred divisors?
module Euler12 = 
    
    // Works but is very slow ~20mins on i5 @ 4.2Ghz
    let numDivisors = 
        
        // triangle number seed 
        let mutable triangleNumberSeed = 1
        
        // A triangle number.
        let mutable triangleNumber = 1
        
        // Number of divisors.
        let mutable numDivisors = 0
        
        // Loop int +1 and check number of divisors until condition.
        while numDivisors <= 500 do
        
            // Reset numDivisors & Triangle num
            triangleNumber <- 0
            numDivisors <- 0
        
            // Generate triangle num
            for i = 1 to triangleNumberSeed do
                triangleNumber <- triangleNumber + i

            // Count divisors of triangle number.
            for i = 1 to triangleNumber do
                if triangleNumber % i = 0 then
                    numDivisors <- numDivisors + 1

            // Increment triangle num.
            triangleNumberSeed <- triangleNumberSeed + 1
        
        // Return triangle num with over 500 divisors.
        printfn "triangle number is %d, num. divisors is %d" triangleNumber numDivisors
        numDivisors